docs: Module 3 Queries, Events - code sample integration (B9lab updates 2-5-a) (#850)
cosmos/sdk-tutorials: b9lab-content/3-main-concepts/13-events.md

## Structure

* A `type` to categorize the Event at a high-level; for example, the Cosmos SDK uses the `message` _type_ to filter events by Msgs.
* A list of `attributes` are key-value pairs that give more information about the categorized event. For example, for the "message" type, we can filter events by key-value pairs using `message.action={some_action}, message.module={some_module}` or `message.sender={a_sender}`.

<HighlightBox type=”info”>

To parse the attribute values as strings, make sure to add `'` (single quotes) around each attribute value.

</HighlightBox>

Events, their type and attributes are defined on a per-module basis in the module's `/types/events.go` file, and are triggered from the module's Protobuf Msg service by using the `EventManager`. Additionally, each module documents its events under `spec/xx_events.md`.
Events are returned to the underlying consensus engine in response to the following ABCI messages:

* `BeginBlock`
* `EndBlock`
* `CheckTx`
* `DeliverTx`

Events are managed by an abstraction called the `EventManager`.

## EventManager

> `EventManager` tracks a list of events for the entire execution flow of a transaction or `BeginBlock`/`EndBlock`. `EventManager` implements a simple wrapper around a slice of `Event` objects that can be emitted from and provides useful methods. The most used method for Cosmos SDK module and application developers is `EmitEvent`.

Reviewer: I am going with:

> For each of the 4 ABCI calls, `CheckTx`, `DeliverTx`, `BeginBlock` and `EndBlock`, the `EventManager` keeps a list of events and delivers it in full as part of the ABCI response. After delivery in an ABCI response, the `EventManager` empties its list of events.

> `EventManager` tracks a list of events for the entire execution flow of a transaction or `BeginBlock`/`EndBlock`. `EventManager` implements a simple wrapper around a slice of `Event` objects that can be emitted from and provides useful methods. The most used method for Cosmos SDK module and application developers is `EmitEvent`.

Reviewer 2: Looks good to me, thanks!

        switch msg := msg.(type) {
    // event types
```

</HighlightBox>

## Subscribing to Events

You can use Tendermint's Websocket (opens new window) to subscribe to events by calling the subscribe RPC method.

The main `eventCategories` you can subscribe to are:

* **NewBlock**: Contains Events triggered during `BeginBlock` and `EndBlock`.
* **Tx**: Contains events triggered during `DeliverTx` (i.e. transaction processing).
* **ValidatorSetUpdates**: Contains validator set updates for the block.

<HighlightBox type=”info”>

→ full list of event categories [here](https://godoc.org/github.com/tendermint/tendermint/types#pkg-constants)

</HighlightBox>

You can filter for event types and attribute values. For example, a transfer transaction triggers an event of type `Transfer` and has `Recipient` and `Sender` as attributes (as defined in the `events.go` file of the bank module.

> 

Reviewer 2: Not a priority, but a wrap up and CTA would be desirable.

Reviewer: Updated.

<ExpansionPanel title="Show me some code for my checkers blockchain">

In your checkers blockchain, it would be good to document a game's lifecycle via events. For instance, when creating the game, you can emit a specific event such that:

```go
var ctx sdk.Context
ctx.EventManager().EmitEvent(
    sdk.NewEvent(sdk.EventTypeMessage,
        sdk.NewAttribute(sdk.AttributeKeyModule, "checkers"),
        sdk.NewAttribute(sdk.AttributeKeyAction, "NewGameCreated"),
        sdk.NewAttribute("Creator", msg.Creator),
        sdk.NewAttribute("Index", newIndex),
        sdk.NewAttribute("Red", msg.Red),
        sdk.NewAttribute("Black", msg.Black),
        sdk.NewAttribute("Wager", strconv.FormatUint(msg.Wager, 10)),
        sdk.NewAttribute("Token", msg.Token),
    ),
)
```
From this model, it is easy to add events to the other transaction types, where you keep in mind that the events are meant to inform and notify relevant parties.

> You ought to also emit an event for games that have timed out. After all, this is part of their lifecycle. You would do it in the end blocker:

Reviewer 2: ```suggestion
You should also emit an event for games that have timed out. After all, this is part of their lifecycle. You would do that in the end blocker:
```
